proposals/010-workflow-migrations.md

```diff
@@ -0,0 +1,303 @@
+```
+
+After registering the metadata object, the lh server will return the object itself.
+
+The following is the wfRunMigrationPlan proto:
+
+```
+message WfRunMigrationPlan {
+    // Name
+    MigrationPlanId id = 1;
+
+    // When the MigrationPlan was created.
+    google.protobuf.Timestamp created_at = 2; 
+
+    // Wrapper object that stores how to migrate every thread
+    MigrationPlan migration_plan = 3 ;
+}
+```
+The first two fields are self-explanatory: the ID used to identify the plan and the time it was created. The third field is a wrapper class where metadata about the migration begins.
+
+The `MigrationPlan` consists of the following proto:
+
+```
+message MigrationPlan { 
+    // {Threadname: {How to migrate that thread}}
+    map<string, ThreadMigrationPlan> thread_migrations = 1;
+}
+```
+Migration plans are broken down to a thread by thread basis.
+Every thread gets its own individual migration plan.
+
+
+```
+// How to migrate a thread 
+message ThreadMigrationPlan {
+    // Name of the new thread we are migrating to
+    string new_thread_name = 1;
+    // name of curNode - > how to migrate curNode in new wfSpec
+    map<string, NodeMigrationPlan> node_migrations  = 2;
+    // Any variable that will be derived before migration node can be provided here
+    // if not provided then possible errors will be thrown
+    map<string, VariableValue> migration_variables = 3;
+}
+```
+
+
+
+For every thread we are migrating we need to know its exact destination in the new `wfSpec`so every thread migration has a new_thread_name -- name of the thread in the new `wfSpec`.
+
+Additionally we need to know what node within the new thread we are migrating to and for that reason we have node_migrations.
+
+migration_variables allows users to inject variableValues into the new Spec. This usefull when variableValues would have been derived prior to the migration node 
+
+```
+//  How to migrate Node 
+message NodeMigrationPlan {
+
+    // Currently just migrate to node with this name 
+    string new_node = 1;
+}
+```
+As of now just the new_node name will be sufficient. A check will be done making sure nodes are of same sub type before migrating.
+
+### Loose `WfRunMigrationPlan` Criteria
+
+This implementation provides loose enforcement of rules when registering the `wfRunMigrationPlan` metadata object, since at registration time we do not yet know the runtime state of any `wfRun` that will be migrated or the version it will migrate to.
+
+This design allows users to define a single `wfRunMigrationPlan` object and apply that migration to multiple `wfRuns` without having to redefine the migration plan for each migration.
+
+### Additional Proto changes
+
+```
+enum GetableClassEnum {
+    WF_RUN_MIGRATION_PLAN = 21;
+}
+```
+
+Getting Object:
+
+```
+  // Get Migration Plan
+  rpc getMigrationPlan(MigrationPlanId) returns (WfRunMigrationPlan) {}
+```
+
+```
+message MigrationPlanId {
+    string name = 1;
```

Comment:
I think I can see why you're not using the `wf_spec_name` or `wf_spec_id` here...is that because this proposal supports defining multiple `MigrationPlan`s that would all move a `WfRun` from `my-workflow` version 0.0 to `my-workflow` version 0.1?